Source/compositorclient/RPI/ModeSet.cpp

```diff
@@ -0,0 +1,392 @@
+                case DRM_NODE_RENDER    :   // Solely for render clients, unprivileged
+                                            {
+                                                if ((1 << type) == (devices[i]->available_nodes & (1 << type)))
+                                                {
+                                                    list.push_back( std::string(devices[i]->nodes[type]) );
+                                                }
+                                                break;
+                                            }
+                case DRM_NODE_MAX       :
+                default                 :   // Unknown (new) node type
+                                        ;
+            }
+        }
+
+        drmFreeDevices(&devices[0], device_count);
+    }
+}
+
+static int FileDescriptor()
+{
+    int fd = -1;
+
+    std::vector<std::string> nodes;
+    GetNodes(DRM_NODE_PRIMARY, nodes);
+
+    std::vector<std::string>::iterator index(nodes.begin());
+
+    while ((index != nodes.end()) && (fd == -1)) {
+        // Select the first from the list
+        if (index->empty() == false)
+        {
+            // The node might be priviliged and the call will fail.
+            // Do not close fd with exec functions! No O_CLOEXEC!
+            fd = open(index->c_str(), O_RDWR); 
+        }
+        index++;
+    }
+    return (fd);
+}
+
+static uint32_t GetConnectors(int fd, uint32_t type)
+{
+    uint32_t bitmask = 0;
+
+    drmModeResPtr resources = drmModeGetResources(fd);
+
+    if(nullptr != resources)
+    {
+        for (int i = 0; i < resources->count_connectors; i++)
+        {
+            drmModeConnectorPtr connector = drmModeGetConnector(fd, resources->connectors[i]);
+
+            if(nullptr != connector)
+            {
+                if ((type == connector->connector_type) && (DRM_MODE_CONNECTED == connector->connection))
+                {
+                    bitmask = bitmask | (1 << i);
+                }
+
+                drmModeFreeConnector(connector);
+            }
+        }
+
+        drmModeFreeResources(resources);
+    }
+
+    return bitmask;
+}
+
+static uint32_t GetCRTCS(int fd, bool valid)
+{
+    uint32_t bitmask = 0;
+
+    drmModeResPtr resources = drmModeGetResources(fd);
+
+    if(nullptr != resources)
+    {
+        for(int i = 0; i < resources->count_crtcs; i++)
+        {
+            drmModeCrtcPtr crtc = drmModeGetCrtc(fd, resources->crtcs[i]);
+
+            if(nullptr != crtc)
+            {
+                if(valid == (1 == crtc->mode_valid ? true : false))
+                {
+                    bitmask = bitmask | (1 << i);
+                }
+            }
+
+            drmModeFreeCrtc(crtc);
+        }
+
+        drmModeFreeResources(resources);
+    }
+
+    return bitmask;
+}
+
+static bool FindProperDisplay(int fd, uint32_t& crtc, uint32_t& encoder, uint32_t& connector, uint32_t& fb)
+{
+    bool found = false;
+
+    assert(fd != -1);
```

Comment:
There is currently no dependency for this source with Thunder. Think it is a bit of overkill to drag it in for the ASSERT, otherwise, would totally agree.